app/bridge/client.js

```diff
@@ -90,85 +121,142 @@ function createNewClient(name, version) {
     this.client = client;
     this.uuid = uuid();
 
-    this.server = null;
+    this.serverChannel = null;
+
+    this.connected = false;
+    this.connectionDeferred = null;
+    this.disconnectionDeferred = null;
+
     this.connect();
   }
 
   ClientInternal.prototype.connect = function() {
-    new Registration(this.client, this.uuid);
-    this.server = new BroadcastChannel(this.uuid);
-    this.listen();
+    debug(this.client.name, this.uuid + ' [connect]');
+
+    if (this.connected) {
+      return Promise.resolve();
+    } else if (this.disconnectionDeferred) {
+      // wait for complete disconnection before trying to connect
+      return this.disconnectionDeferred.promise.then(() => this.connect());
+    } else if (this.connectionDeferred) {
+      return this.connectionDeferred.promise;
+    } else {
+      this.connectionDeferred = new Deferred();
+      register(this.client, this.uuid);
+      // It might not be the first time we connect
+      if (!this.serverChannel) {
+        this.serverChannel = new BroadcastChannel(this.uuid);
+        this.listen();
+      }
+      return this.connectionDeferred.promise;
+    }
   };
 
   ClientInternal.prototype.onconnected = function(contract) {
-    debug(this.client.name + ' [connected]');
+    debug(this.client.name, this.uuid, ' [connected]');
+    // we might not be waiting for connection
+    if (this.connectionDeferred) {
+      this.connectionDeferred.resolve();
+    }
+    this.connectionDeferred = null;
 
-    this.connected = true;
+    if (!this.connected) {
```

Comment:
I figured that I shouldn't trust other threads to send me one and only one `connected` message. Also we have asynchronous message. We can receive multiple `disconnected` and `connecteded` events before having time to even run the first one (for example if you go to background immediately after starting the app, then go back to foreground). In my experience, being extra defensive here helps a lot to have robust behaviour. Maybe I'm extra-cautious here though.